### PR TITLE
feat: UI Resubmit workflows with parameter (#4662)

### LINKED ISSUE #4662: Need ability to change parameters/arguments while submitting/re-submitting workflows in Argo UI
# Summary

Need ability to change parameters/arguments while resubmitting workflows in Argo UI

# Use Case

In case of failures or if we want to run the workflow again with different parameters/arguments, we need ability to do so using Argo UI

### PR DESCRIPTION
**Pull request by toyamagu-2021**
Signed-off-by: toyamagu2021@gmail.com <toyamagu2021@gmail.com>

Fixes #4662

### Motivation

- We can change parameters when submitting [archived] workflows.

### Modifications

- Popup side panel to change parameters when we click resubmit button in [archived] workflows details page.

#### To check regression

<details>
<summary>manifest</summary>

```yaml
spec:
  templates:
    - name: main
      inputs: {}
      outputs: {}
      metadata: {}
      dag:
        tasks:
          - name: step-a
            template: step-template-a
            arguments:
              parameters:
                - name: template-param-1
                  value: '1'
                - name: template-param-2
                  value: 'yes'
                  enum:
                    - 'yes'
                    - 'no'
          - name: step-b
            template: step-template-b
            arguments: {}
    - name: step-template-a
      inputs:
        parameters:
          - name: template-param-1
          - name: template-param-2
            enum:
              - 'yes'
              - 'no'
      outputs: {}
      metadata: {}
      script:
        name: ''
        image: alpine
        command:
          - /bin/sh
        resources: {}
        source: |
          echo "{{inputs.parameters.template-param-1}}"
          echo "{{inputs.parameters.template-param-2}}"
    - name: step-template-b
      inputs:
        parameters:
          - name: workflow-param-1
            value: '{{workflow.parameters.workflow-param-1}}'
          - name: workflow-param-2
            value: '{{workflow.parameters.workflow-param-2}}'
      outputs: {}
      metadata: {}
      container:
        name: main
        image: argoproj/argosay:v2
        command:
          - /argosay
        args:
          - echo
          - >-
            {{inputs.parameters.workflow-param-1}}-{{inputs.parameters.workflow-param-2}}
        resources: {}
  entrypoint: main
  arguments:
    parameters:
      - name: workflow-param-2
      - name: workflow-param-1
        value: 'yes'
        enum:
          - 'yes'
          - 'no'
```

</details>

<details>
<summary>Submit workflows</summary>

- image
- image
- image

</details>

<details>
<summary>Submit workflows from workflow templates select entrypoint</summary>

- image
- image

</details>

#### Check to resubmit workflow

[manifest](https://github.com/argoproj/argo-workflows/pull/11083/files#diff-e8056c5ea3d9eeb2bc56ed0809defcaa1de7f666735136b4edfcd0bdd8f96769)

<details>
<summary>Resubmit to success</summary>

- before
  image
- resubmit
  image
- after
  image

</details>

<details>
<summary>Resubmit to fail</summary>

- before:
  image
- resubmit
  image
- after
  image

</details>

<details>
<summary>Resubmit to success with memoized</summary>

- before
  image
- resubmit
  image
- after
  image

</details>

<details>
<summary>Resubmit archived workflow</summary>

- before (Note that `Pending` status means workflow is deleted by GC)
  image
- resubmit
  image
- after
  image
</details>

### CODE REVIEW

File: ui/src/app/workflows/components/resubmit-workflow-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                            <>
+                                <br />
+                                <label>No parameters</label>
+                            </>
+                        ) : (
+                            <></>
+                        )}
+                    </div>
+
+                    <div key='memorized' style={{marginBottom: 25}}>
```
Comment (terrytangyuan) [inline ui/src/app/workflows/components/resubmit-workflow-panel.tsx:58]: Change to "memoized" to be consistent?

File: ui/test/ui/ui-memoize.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+apiVersion: argoproj.io/v1alpha1
+kind: WorkflowTemplae
```
[the remark this answers is not in the document]
Comment (toyamagu-2021) [reply ui/test/ui/ui-memoize.yaml:2]: Yes, I used this to do manual UI tests.
But I had copied and pasted `spec` field only...

**Comment by toyamagu-2021** [comment]
Flaky tests are failing, but I'd like to get community feedback before making these successes.

**Comment by terrytangyuan** [comment]
I am working on https://github.com/argoproj/argo-workflows/pull/11121 so this should be implemented in workflows detail/list page once that's merged.

**Comment by toyamagu-2021** [comment]
Fair enough!
I' will rebase these changes after your PR is merged.

File: ui/src/app/workflows/components/submit-workflow-panel.tsx [hunk 3]
```diff
@@ -115,81 +118,14 @@ export class SubmitWorkflowPanel extends React.Component<Props, State> {
...
-            return p.default;
-        } else {
-            return p.value;
-        }
-    }
-
     private submit() {
         this.setState({isSubmitting: true});
         services.workflows
             .submit(this.props.kind, this.props.name, this.props.namespace, {
```
Comment (toyamagu-2021) [inline ui/src/app/workflows/components/submit-workflow-panel.tsx:124]: We can use an endpoint for Archived workflows here, but I think not necessary.

https://github.com/toyamagu-2021/argo-workflows/blob/df85d398af7596f99fcfcfe53688e5edbf1d2f79/server/workflow/workflow_server.go#L634

File: ui/src/app/workflows/components/resubmit-workflow-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        const parameters: ResubmitOpts['parameters'] = this.state.overrideParameters
+            ? [...this.state.workflowParameters.filter(p => Utils.getValueFromParameter(p) !== undefined).map(p => p.name + '=' + Utils.getValueFromParameter(p))]
+            : [];
+        const opts: ResubmitOpts = {
+            parameters,
+            memoized: this.state.memoized
+        };
+
+        services.workflows
+            .resubmit(this.props.workflow.metadata.name, this.props.workflow.metadata.namespace, opts)
```
Comment (terrytangyuan) [inline ui/src/app/workflows/components/resubmit-workflow-panel.tsx:102]: Could you check if this works for archived workflow as well?

Comment (toyamagu-2021) [reply ui/src/app/workflows/components/resubmit-workflow-panel.tsx:102]: This works basically (please see PR description), but I noticed an edge case where this logic will not work where there is more than one workflow that has the same name and namespace.
Could you comment on this issue #11371?

**Review by terrytangyuan — APPROVED** [review]
This one looks good. We can discuss remaining issue separately.

**Pull request merged**